tokio-fs/src/file/mod.rs

```diff
@@ -121,6 +123,11 @@ impl File {
         ::blocking_io(|| self.std().set_len(size))
     }
 
+    /// Queries metadata about the underlying file.
+    pub fn metadata(self) -> MetadataFuture {
```

Comment:
It makes very little sense to me to take `self` here, if you have to metadata wouldn't want to use the `File` some more? `&mut self` would make more sense to me.

Reply:
> It makes very little sense to me to take self here, if you have to metadata wouldn't want to use the File some more?

Yes, you would, which is why I'm submitting this PR.

It's basically impossible to have both the file and the metadata when taking a reference. Doing as you suggest and taking a `&mut self` yields this type of usage:

```rust
        File::create(file_path)
            .and_then(|mut file| file.metadata()})
            .and_then( /* ... */ )
```

How do you pass the `file` along so that it's usable in the last `and_then`? You can't move it out of the first `and_then` because the pending `Metadata` future has a reference. Additionally, when `file` goes out of scope at the end of the closure, it also means that the reference held by the pending closure would be invalidated.

I would welcome your suggested changes, but I cannot figure out how to make them work. In fact, using `&mut self` was my original attempt when using `poll_metadata`, and @carllerche told me to write this and submit it.

I encourage you to pull down this PR and give your suggested changes a whirl; I look forward to learning how it can be done the way you say it can be.

Reply:
I wasn't thinking about using `and_then`, I thinking of using `await`, along the following lines:

```rust
let file = /* ... */
let metadata = await file.metadata();
// Continue with both.
```

But using `and_then` would likely make that somewhat awkward to use.

Reply:
> I thinking of using `await`

Ah. Some of us use futures in stable Rust, so `await` isn't an option. I wonder how that ultimately desugars...